doc/advanced/content/executor_design.rst

```diff
@@ -0,0 +1,421 @@
+   struct inline_executor {
+
+      // Shape and index are specified to the custom PCL type uindex_t
+      using shape_type = uindex_t ;
+      using index_type = uindex_t ;
+
+
+      // execute invokes the callable exactly once
+      template <typename F>
+      void execute(F&& f) const {
+        f();
+      }
+
+      // bulk_execute invokes the callable n times and passes the index of execution agent
+      // In case of inline_executor each execution agent is mapped to the same thread is invoked sequentially
+      template <typename F, typename... Args>
+      void bulk_execute(F&& f, const shape_type& n) const {
+        for (shape_type index = 0; index <  n; ++index)
+          f(index);
+      }
+
+      // The query customization point being specialized to show that the executor supports
+      // the blocking.always property
+      static constexpr auto query(const blocking_t&) noexcept {
+        return blocking_t::always;
+      }
+
+      // The require customization point being specialized to allow strongly associating the executor with
+      // the blocking.always property
+      inline_executor<blocking_t::always_t, ProtoAllocator> require(const blocking_t::always_t&) const {
+        return {};
+      }
+    };
+
+
+Best Fit Executor
+=================================
+
+In most scenarios users of PCL do not care much about performance and simply want their code to run
+fast without any additional steps performed from their side. For this reason their was a need for
+mechanism called best fit, which automatically selects best possible executor and its customizes its
+property in order to give the user good performance out of the box based on their system configurations
+and the function they are calling.
+
+The mechanism works be choosing an executor based on the availability of an executor (depends on
+hardware/software of a system) and the priority of executors specified in function by PCL
+maintainers & contributors which will give good performance. The executor properties can also be
+customized on for a better fit for certain scenarios. Besides this runtime checks are also be specified
+and on the basis of these runtime checks the executors are further filtered to selected most appropriate
+one. The two supported mechanisms in PCL currently are `enable_exec_with_priority` and
+`enable_exec_on_desc_priority`. You can read more about them in the code API.
+
+Alternative Designs Considered
+=================================
+
+There has been a lot of deliberation and discussion regarding all the design aspects of executors in PCL.
+The design has gone through multiple iterations before a consensus was reached. Some of the major alternate
+design proposals that were rejected were:
+
+* Tag Dispatching
+
+This was was one of the initial design considerations. It seemed like an attractive choice due to simplicity,
+could choose the allowed overload resolution to choose the best option at compile time by allowing
```

Comment:
due to its simplicity, allowing overload resolution to choose the best options at compile time.